app/controllers/application_controller.rb

```diff
@@ -254,12 +254,18 @@ def pull_pinned_links_and_hot_questions
         PinnedLink.where(active: true).where('shown_before IS NULL OR shown_before > NOW()').all
       end
     end
+
+    # We need to filter out pinned links that aren't posts
+    # because for whatever reason, where.not(id: nil) will return an empty query
+    pinned_post_ids = @pinned_links.pluck(:post_id).select(&:present?)
+  
     @hot_questions = Rails.cache.fetch('hot_questions', expires_in: 4.hours) do
       Rack::MiniProfiler.step 'hot_questions: cache miss' do
         Post.undeleted.where(last_activity: (Rails.env.development? ? 365 : 7).days.ago..DateTime.now)
             .where(post_type_id: [Question.post_type_id, Article.post_type_id])
             .joins(:category).where(categories: { use_for_hot_posts: true })
             .where('score >= ?', SiteSetting['HotPostsScoreThreshold'])
+            .where.not(id: pinned_post_ids)
```

Comment:
I have not tested this, but to optimize the query it would be better if we do not pluck.

Also, your finding of pinned links which do not have a post id breaking the query is quite strange. Especially since "WHERE id IS NOT NULL" should not affect the query.

```suggestion
            .where.not(id: @pinned_links.select(:post_id))
```

Reply:
Unfortunately, that doesn't work, and I really have no idea why.

The debugger shows

![image](https://user-images.githubusercontent.com/54333972/187692576-598269a3-a7cc-4aa2-a962-2d3f36a04e19.png)

```
[#<PinnedLink id: nil, post_id: 28>, #<PinnedLink id: nil, post_id: nil>, #<PinnedLink id: nil, post_id: 21>]
```

(Of course, I didn't add the to_a in the code itself, it was just so I could see what came up)

Reply:
Thanks for checking. I had a look at `@pinned_posts` and it seems to be cached in redis, which may be what causes the interesting behavior. I think then your solution is the best one indeed.

Reply:
So yes, filtering out the null values from the featured links is unfortunately necessary

Reply:
Sigh. Could you drop a comment in there for the next person who comes across this code and thinks it can be simplified?

Reply:
@cellio Done 👍